containers/src/Data/Graph.hs

```diff
@@ -139,15 +151,26 @@ default ()
 -------------------------------------------------------------------------
 
 -- | Strongly connected component.
-data SCC vertex = AcyclicSCC vertex     -- ^ A single vertex that is not
-                                        -- in any cycle.
-                | CyclicSCC  [vertex]   -- ^ A maximal set of mutually
-                                        -- reachable vertices.
+data SCC vertex
+  = AcyclicSCC vertex
+  -- ^ A single vertex that is not in any cycle.
+  | NECyclicSCC {-# UNPACK #-} !(NonEmpty vertex)
+  -- ^ A maximal set of mutually reachable vertices.
+  --
+  -- @since 0.7.0
   deriving ( Eq   -- ^ @since 0.5.9
            , Show -- ^ @since 0.5.9
            , Read -- ^ @since 0.5.9
            )
 
+-- | Partial pattern synonym for backward compatibility with @containers < 0.7@.
+pattern CyclicSCC :: [vertex] -> SCC vertex
```

Comment:
Doesn't the pattern synonym definition also need to be guarded by `#ifdef __GLASGOW_HASKELL__` ? 
I'm always confused by the portability story. Tangentially related, how do we even test this given CI only uses ghc?

Reply:
Yes it does. The portability story kind of sucks right now. The main goal as I see it is to make sure it won't be too big a pain to add support for a future Report-style Haskell (Purescript, sadly, is just a bit too far away to be a practical target for this). I think we should stop using `__GLASGOW_HASKELL__` directly and instead use a few of our own macros. We should have a build job that pretends were not on GHC; getting that to do testing right will take a bit of fussing, but it's *probably* doable?